Content.Shared/GameObjects/Components/Chemistry/SharedSolutionContainerComponent.cs

```diff
@@ -1,115 +1,228 @@
-            set
-            {
-                if (_maxVolume == value)
-                {
-                    return;
-                }
+            base.ExposeData(serializer);
 
-                _maxVolume = value;
-                Dirty();
-            }
+            serializer.DataField(this, x => x.CanReact, "canReact", true);
+            serializer.DataField(this, x => x.MaxVolume, "maxVol", ReagentUnit.New(0));
+            serializer.DataField(this, x => x.Solution, "contents", new Solution());
+            serializer.DataField(this, x => x.Capabilities, "caps", SolutionContainerCaps.AddTo | SolutionContainerCaps.RemoveFrom | SolutionContainerCaps.CanExamine);
         }
 
-        /// <summary>
-        ///     The current blended color of all the reagents in the container.
-        /// </summary>
-        [ViewVariables(VVAccess.ReadWrite)]
-        public virtual Color SubstanceColor
+        public void RemoveAllSolution()
+        {
+            if (CurrentVolume > 0)
+                return;
+
+            Solution.RemoveAllSolution();
+            ChemicalsRemoved();
+        }
+
+        public bool TryAddReagent(string reagentId, ReagentUnit quantity, out ReagentUnit acceptedQuantity)
+        {
+            acceptedQuantity = EmptyVolume > quantity ? quantity : EmptyVolume;
+            Solution.AddReagent(reagentId, acceptedQuantity);
+
+            if (acceptedQuantity > 0)
+                ChemicalsAdded();
+
+            return acceptedQuantity == quantity;
+        }
+
+        public bool TryRemoveReagent(string reagentId, ReagentUnit quantity)
+        {
+            if (!Solution.ContainsReagent(reagentId))
+                return false;
+
+            Solution.RemoveReagent(reagentId, quantity);
+            ChemicalsRemoved();
+            return true;
+        }
+
+        public Solution SplitSolution(ReagentUnit quantity)
+        {
+            var splitSol = Solution.SplitSolution(quantity);
+            ChemicalsRemoved();
+            return splitSol;
+        }
+
+        public bool CanAddSolution(Solution solution)
+        {
+            return solution.TotalVolume <= EmptyVolume;
+        }
+
+        public bool TryAddSolution(Solution solution)
+        {
+            if (!CanAddSolution(solution))
+                return false;
+
+            Solution.AddSolution(solution);
+            ChemicalsAdded();
+            return true;
+        }
+
+        private void ChemicalsAdded()
+        {
+            ProcessReactions();
+            SolutionChanged();
+            UpdateAppearance();
+            Dirty();
+        }
+
+        private void ChemicalsRemoved()
+        {
+            SolutionChanged();
+            UpdateAppearance();
+            Dirty();
+        }
+
+        private void SolutionChanged()
+        {
+            IoCManager.Resolve<IEntitySystemManager>()
+                .GetEntitySystem<ChemistrySystem>()
+                .HandleSolutionChange(Owner);
```

Comment:
```suggestion
            EntitySystem
                .Get<ChemistrySystem>()
                .HandleSolutionChange(Owner);
```

If you wanted you could also use the eventbus and use QueueEvent so it's processed at the end of a tick.

Reply:
I'll look into QueueEvent in the future, not currently sure if it will break any functionality to delay the method call.